### PR TITLE
[server] abort running prebuilds on same branch

### LINKED ISSUE #6569: Pushing new commits to a branch, does not stop a prebuild from a previous commit of the same branch.
### Bug description

I think a previously running prebuild should be cancelled if new commits are pushed to the same branch. 

I noticed it after I pushed changes to a branch, and realized I forgot to make a small change.
image

### Steps to reproduce

Create long init task in `.gitpod.yml`.
After pushing a commit that will trigger the prebuild, push another commit to the same branch, and go to Prebuilds page on your gitpod.io account.

### Workspace affected

_No response_

### Expected behavior

Seems like waste of resources to keep the previous prebuild running.
Not sure if this might cause other issues (like an old prebuild finish running after the latest prebuild, and perhaps override things?)

### Example repository

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
Cancels all running and queued prebuilds for the same branch when a new commit is pushed. 
Adds a project setting to opt-out of this behavior.

Screenshot 2022-06-28 at 12 54 36

## Related Issue(s)

Fixes #6569

## How to test
- Add a project with a long prebuild task
- push multipüle times to the same branch
- see how the previous prebuilds get cancelled.
- push to a different branch and verify that running prebuilds on the other branch are not getting cancelled.
- disable the project setting and verify that the cancellation doesn't happen

## Release Notes

```release-note
Outdated prebuilds (i.e. new commits are pushed on a branch) are automatically canceled. This behavior can be disabled in the project's settings.
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by jankeromnes** [comment]
Many thanks @svenefftinge for continuously improving & optimizing which prebuilds we should be running or not! 💯

I'll give this Pull Request a look. 👀

File: components/dashboard/src/projects/ProjectSettings.tsx [hunk 1]
```diff
@@ -56,19 +55,14 @@ export default function () {
...
 
     useEffect(() => {
         if (!project) {
             return;
         }
-        setIsLoading(false);
-        setIsIncrementalPrebuildsEnabled(!!project.settings?.useIncrementalPrebuilds);
-        setIsPersistentVolumeClaimEnabled(!!project.settings?.usePersistentVolumeClaim);
-
+        setProjectSettings({ ...project.settings });
```
Comment (jankeromnes) [inline components/dashboard/src/projects/ProjectSettings.tsx:65]: What is the benefit of having an extra `projectSettings` state here, given that we already have a `project`? 🤔

Couldn't we use `project.settings` as the source of truth? (And modify that + call `updateProjectSettings` when things are toggled?)

Comment (svenefftinge) [reply components/dashboard/src/projects/ProjectSettings.tsx:65]: Because we are using react state and need to update the object for react to rerender. react doesn't check deep references

File: components/dashboard/src/projects/ProjectSettings.tsx [hunk 2]
```diff
@@ -56,19 +55,14 @@ export default function () {
     const { teams } = useContext(TeamsContext);
     const team = getCurrentTeam(location, teams);
 
-    const [isLoading, setIsLoading] = useState<boolean>(true);
```
Comment (jankeromnes) [inline components/dashboard/src/projects/ProjectSettings.tsx:59]: ❤️ Many thanks for removing a `isLoading` boolean! (which I might have added 🙈)

I agree that maintaining a `isLoading` boolean is generally bad practice, and creates many opportunities for bugs. (And when a loading state is required, it's always better to explicit wait on what still needs to be loaded, instead of trying to code this into a `isLoading` boolean.) 👍

File: components/dashboard/src/projects/ProjectSettings.tsx [hunk 3]
```diff
@@ -146,10 +125,15 @@ export default function () {
                         </a>
                     </span>
                 }
-                checked={isIncrementalPrebuildsEnabled}
-                disabled={isLoading}
```
Comment (jankeromnes) [inline components/dashboard/src/projects/ProjectSettings.tsx:150]: I wonder if we should keep the disabled state of checkboxes, e.g. for when `project` isn't loaded yet? 🤔 This could avoid having interactive checkboxes in the wrong state (e.g. everything seems disabled while we're loading the actual project settings.)

Not super important though. 👍

File: components/dashboard/src/projects/ProjectSettings.tsx [hunk 4]
```diff
@@ -146,10 +125,15 @@ export default function () {
...
+                checked={!!projectSettings.useIncrementalPrebuilds}
                 onChange={toggleIncrementalPrebuilds}
             />
+            <CheckBox
+                title={<span>Cancel Prebuilds on Outdated Commits </span>}
+                desc={<span>Cancels all pending and running prebuilds on the branch when a new commit is pushed.</span>}
```
Comment (jankeromnes) [inline components/dashboard/src/projects/ProjectSettings.tsx:133]: In general, I think that this feature is super valuable and needed, and I'm happy that we're taking a first step towards it. 💯

Without having read the code much further or tested it, one initial thought that occurs is that, if one of your branches (e.g. `main`) takes a long time to build, and sees very frequent commits (e.g. during working hours), we might be always-cancelling running prebuilds without ever finishing one, for long stretches of time (e.g. maybe we only successfully finish running a prebuild over lunch break or at the end of the day).

File: components/dashboard/src/projects/ProjectSettings.tsx (hunk 1 above)
Comment (jankeromnes) [reply components/dashboard/src/projects/ProjectSettings.tsx:65]: Ah, that's a good point, thanks.

So I guess getting `setProject` from the ProjectContext, and calling `setProject(project)` after modifying the `project.settings` wouldn't actually update the state, right? 🤔

In any case, the current implementation is fine by me assuming it works 😊 (will test it soon) so this is not a blocker.

File: components/server/ee/src/prebuilds/prebuild-manager.ts [hunk 5]
```diff
@@ -58,6 +58,22 @@ export class PrebuildManager {
...
 
+    async abortPrebuildsForBranch(ctx: TraceContext, project: Project, user: User, branch: string): Promise<void> {
+        const span = TraceContext.startSpan("abortPrebuildsForBranch", ctx);
+        const prebuilds = await this.workspaceDB
+            .trace({ span })
+            .findActivePrebuiltWorkspacesByBranch(project.id, branch);
+        const results: Promise<any>[] = [];
+        for (const prebuild of prebuilds) {
+            for (const instance of prebuild.instances) {
+                results.push(this.workspaceStarter.stopWorkspaceInstance({ span }, instance.id, instance.region));
```
Comment (jankeromnes) [inline components/server/ee/src/prebuilds/prebuild-manager.ts:69]: I wonder what happens for instances which are not in a "stoppable" state, e.g. in image build? Will this fail for some not-yet-stoppable instances, or will this wait until it reaches a stoppable state? (I'm guessing the former.)

File: components/server/src/workspace/workspace-starter.ts [hunk 6]
```diff
@@ -349,6 +351,20 @@ export class WorkspaceStarter {
         }
     }
 
+    public async stopWorkspaceInstance(
```
Comment (jankeromnes) [inline components/server/src/workspace/workspace-starter.ts:354]: Good idea moving this here. 👍

File: components/server/src/workspace/workspace-starter.ts [hunk 7]
```diff
@@ -349,6 +351,20 @@ export class WorkspaceStarter {
...
+        ctx: TraceContext,
+        instanceId: string,
+        instanceRegion: string,
+        policy?: StopWorkspacePolicy,
+    ): Promise<void> {
+        const req = new StopWorkspaceRequest();
+        req.setId(instanceId);
+        req.setPolicy(policy || StopWorkspacePolicy.NORMALLY);
+
+        const client = await this.clientProvider.get(instanceRegion);
```
Comment (jankeromnes) [inline components/server/src/workspace/workspace-starter.ts:364]: If I remember correctly, some instances might not have a region yet (e.g. still in image build or so?) and this could cause `client` to be incorrect or undefined. Is this still the case? 🤔 I guess this needs to be tested.

**Review by jankeromnes — COMMENTED** [review]
Thanks again! Had an initial pass through the code, and it looks good to me, modulo a few (mostly optional) questions.

Will now give the PR a go, and try to find some edge cases (e.g. image builds?)

**Comment by jankeromnes** [comment]
Err, will actually test this PR later today (in ~1.5 hours) as I have upcoming meetings. Please feel free to push any changes if you want to (i.e. no need to necessarily keep the preview env alive until then).

**Comment by geropl** [comment]
@svenefftinge Do you have any plans on announcing this change? This will certainly affect some heavy Gitpod users. :thinking:

**Comment by svenefftinge** [comment]
> This will certainly affect some heavy Gitpod users. 🤔

How do you think it will affect users? In a negative way you mean?

**Comment by svenefftinge** [comment]
The only path where users might be surprised is seeing prebuilds "CANCELED". We should probably tell them why

File: components/dashboard/src/projects/ProjectSettings.tsx (hunk 3 above)
Comment (svenefftinge) [reply components/dashboard/src/projects/ProjectSettings.tsx:150]: I'd rather work on making loading a single project faster if that's really an issue

File: components/dashboard/src/projects/ProjectSettings.tsx (hunk 1 above)
Comment (svenefftinge) [reply components/dashboard/src/projects/ProjectSettings.tsx:65]: Yes, it does not update the state.

**Comment by jankeromnes** [comment]
>  I'm adding the status descriptions as a tooltip for the items in the list and also add the information about new commits being pushed on the branch:

Good idea! 👍 I wonder if it also makes sense to tell users that they can click on the prebuild in order to know the exact cancellation reason. (It should always be visible in the full error message below the prebuild logs.)

**Comment by svenefftinge** [comment]
> Good idea! 👍 I wonder if it also makes sense to tell users that they can click on the prebuild in order to know the exact cancellation reason. (It should always be visible in the full error message below the prebuild logs.)

It is the same message

**Comment by gtsiolis** [comment]
[FWIW](https://en.wiktionary.org/wiki/FWIW), auto-cancelling redundant builds is a common practice in CI systems[[1](https://docs.gitlab.com/ee/ci/pipelines/settings.html#auto-cancel-redundant-pipelines)][[2](https://circleci.com/docs/2.0/skip-build#auto-cancelling)][...].

In the future, we could introduce a preference option in project settings but going with a sensible default[[3](https://en.wikipedia.org/wiki/Convention_over_configuration)] sounds better.

**Comment by svenefftinge** [comment]
> In the future, we could introduce a preference option in project settings but going with a sensible default[[3](https://en.wikipedia.org/wiki/Convention_over_configuration)] sounds better.

This PR already comes with that preference (default is ☑️ ):
Screenshot 2022-06-29 at 16 36 43

**Review by gtsiolis — APPROVED** [review]
>This PR already comes with that preference (default is ☑️ ) 

Perfect (Perfect)! ✨

Left one comment for updating the copy used in settings. Feel free to ignore or adjust the suggestion. 🏓

File: components/server/ee/src/prebuilds/prebuild-manager.ts [hunk 8]
```diff
@@ -58,6 +58,22 @@ export class PrebuildManager {
...
+            }
+            prebuild.prebuild.state = "aborted";
```
Comment (jankeromnes) [inline components/server/ee/src/prebuilds/prebuild-manager.ts:71]: Please also specify the cancellation reason by setting a helpful `prebuild.error` value, similar to the [rate-limit error](https://github.com/gitpod-io/gitpod/blob/7d005012a82778e5b916bf494a2c64bd461fec7d/components/server/ee/src/prebuilds/prebuild-manager.ts#L188-L191):

```suggestion
            prebuild.prebuild.state = "aborted";
            prebuild.error = "A newer commit was pushed to the same branch.";
```

**Review by jankeromnes — APPROVED** [review]
Many thanks! In addition to the code looking good to me earlier, I've tested this PR and it works as advertised. ✨ 🎯

For what it's worth, I also tested the behavior with image builds (by pushing a custom Dockerile and then several commits on top) -- the Prebuilds actually get cancelled, even though the logs continue to receive new updates. 😅 I guess that's fine. 👍

Screenshot 2022-06-29 at 19 12 24

Adding a hold because I'd love the cancellation reason to be explicitly added as a `prebuild.error`, or in case someone else would like to review.

/hold

**Comment by svenefftinge** [comment]
> I'd love the cancellation reason to be explicitly added as a prebuild.error,

Will do in a follow up PR.
